### PR TITLE
Add documentation and testing for `dimod.typing`

### PR DESCRIPTION
**Pull request by arcondello**
There is more testing needed, but that will come in followup PRs. For now just putting the minimum amount in that's needed to check that the approach works

### CODE REVIEW

File: docs/reference/index.rst [hunk 1]
```diff
@@ -14,6 +14,7 @@ Reference Documentation
    generators
    sampler_composites/index
    sampleset
+   typing
```
Comment (JoelPasvolsky) [inline docs/reference/index.rst:17]: These are not alphabetical, so why not put this after Exceptions? I think users will likely first need utilities and the pages related to variables and their types before they get to this. 
I can live with it here though too.

File: dimod/typing.py [hunk 2]
```diff
@@ -48,67 +46,127 @@
...
+* NumPy's :class:`~numpy.floating` and :class:`~numpy.integer`.
+
+"""
+
+# it would be nice to be able to exclude None from Variable. We can define
+# a special abstract base class that excludes None, but making it behave like
+# a typing class is a pain accross our supported Python versions. In the future
+# we should handle it
+
+Variable = typing.Hashable
```
Comment (JoelPasvolsky) [inline dimod/typing.py:64]: I'm wondering whether the uglier `VariableLabel` would help reduce confusion between QMs representing variables and labels of variables in the models...

File: dimod/vartypes.py [hunk 3]
```diff
@@ -121,7 +121,18 @@ class Vartype(enum.Enum):
 REAL = Vartype.REAL
 
 
+# when we drop 3.7 we can just use typing.Literal
 VartypeLike = typing.Union[Vartype, str, frozenset]
+"""Objects that can be interpreted as a variable type.
```
Comment (JoelPasvolsky) [inline dimod/vartypes.py:126]: This docstring isn't displayed in the typing.rst output HTML

**Review by JoelPasvolsky — APPROVED** [review]
Anything that reduces the typing character count of ``Initialized.parse_initial_states`` from 3000 to 700 is a blessing

File: dimod/typing.py [hunk 4]
```diff
@@ -48,67 +46,127 @@
...
     typing.Mapping[Variable, Bias],
     ArrayLike,  # this is overgenerous, but we cannot easily specify it better
     ]
+"""Objects that can be interpreted as a single sample.
+
+This includes:
+
+* A one-dimensional NumPy array_like_.
+* A 2-:class:`tuple` containing a one-dimensional NumPy array_like_ and a
+  list of variable labels.
```
Comment (randomir) [inline dimod/typing.py:141]: Is this included in the definition above?

File: tests/test_typing.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    TEST_DIR = os.path.join(os.path.dirname(__file__), 'typing')
+    CACHE_DIR = os.path.join(TEST_DIR, '.mypy_cache')
+
+    def test_mypy(self):
+        stdout, stderr, exit_code = mypy.api.run([
+            '--cache-dir', self.CACHE_DIR,
+            '--follow-imports', 'silent',
+            '--show-absolute-path',
+            self.TEST_DIR])
+
```
Comment (randomir) [inline tests/test_typing.py:41]: Perhaps replace this python unittest with a CI job that just calls `mypy`?

**Review by randomir — APPROVED** [review]
LGTM

File: dimod/typing.py (hunk 2 above)
Comment (arcondello) [reply dimod/typing.py:64]: Yes. I had suggested that a while ago as you'll recall. I am open to it, though probably as a followup PR

Comment (arcondello) [reply dimod/typing.py:64]: Made an issue https://github.com/dwavesystems/dimod/issues/1160

File: tests/test_typing.py (hunk 5 above)
Comment (arcondello) [reply tests/test_typing.py:41]: The problem is checking the failure cases. MyPy doesn't really have a very specific exclude syntax.

File: docs/reference/index.rst (hunk 1 above)
Comment (arcondello) [reply docs/reference/index.rst:17]: Sure, can move it.

File: dimod/vartypes.py (hunk 3 above)
Comment (arcondello) [reply dimod/vartypes.py:126]: I am going to leave it alone for now - see what RTD does with it

**Pull request merged**